theano/tensor/opt.py

```diff
@@ -5696,11 +5696,24 @@ def local_opt_alloc(node):
                                                     only_process_constants=True)
                     assert val.size == 1
                     # check which type of op
-                    casted = T.mul(*shapes).astype(str(input.dtype))
+                    size = T.mul(*shapes)
+                    if input.dtype == "float32":
+                        # shapes are ints and normally int64.
+                        # We don't want to have a float64 upcast here
+                        # if input is a float32.
+                        size = size.astype(input.dtype)
```

Comment:
The "if" on line 5692 means we are in a case where the reduction was on all axes. The "else" block after line 5721 is the case where the reduction is only done on some axes.
The code in both blocks is similar, but the the cast `size = size.astype(input.dtype)` only happens in the first block, here, and not in the other block (for instance, just after line 5730).
Why that difference?